packages/protocol/contracts/ACE/ACE.sol

```diff
@@ -186,6 +187,148 @@ contract ACE is IAZTEC {
+
+        updateOutputNotes(mintedNotes);
+        return(newTotal, mintedNotes);
+    }
+
+    /**
+    * @dev Call transferFrom on a linked ERC20 token. Used in cases where the ACE's mint
+    * function is called but the token balance of the note registry in question is
+    * insufficient
+    * @param _value the value to be transferred
+    */
+    function supplementTokens(uint256 _value) external {
+        require(_value > uint(0), "supplement tokens can only be called on a positive value");
+        NoteRegistry storage registry = registries[msg.sender];
+        require(registry.flags.active == true, "note registry does not exist for the given address");
+        require(registry.flags.canConvert == true, "note registry does not have conversion rights");
+        
+        // Only scenario where supplementTokens() should be called is when a mint/burn operation has been executed
+        require(registry.flags.canAdjustSupply == true, "note registry does not have mint and burn rights");
+        
+        require(
+            registry.linkedToken.transferFrom(msg.sender, address(this), _value), 
+            "transfer failed"
+        );
+    }
+
+    /**
+    * @dev Burn AZTEC notes
+    *      
+    * @param _proof the AZTEC proof object
+    * @param _proofData the burn proof construction data
+    * @param _proofSender the Ethereum address of the original transaction sender. It is explicitly assumed that
+    *        an asset using ACE supplies this field correctly - if they don't their asset is vulnerable to front-running
+    * Unnamed param is the AZTEC zero-knowledge proof data
+    * @return two `bytes` objects. The first contains the new confidentialTotalSupply note and the second contains the
+    * notes that were created. Returned so that a zkAsset can emit the appropriate events
+    */
+    function burn(
+        uint24 _proof,
+        bytes calldata _proofData,
+        address _proofSender
+    ) external returns (bytes memory, bytes memory) {
+        
+        NoteRegistry storage registry = registries[msg.sender];
+        require(registry.flags.active == true, "note registry does not exist for the given address");
+        require(registry.flags.canAdjustSupply == true, "this asset is not burnable");
+        
+        // Check that it's a burnable proof
+        (, uint8 category, ) = _proof.getProofComponents();
+
+        require(category == uint8(ProofCategory.BURN), "this is not a burn proof");
+
+        bytes memory _proofOutputs = this.validateProof(_proof, _proofSender, _proofData);
+        
+        // Dealing with notes representing totals
+        (bytes memory oldTotal, // input notes
+        bytes memory newTotal, // output notes
+        ,
+        int256 publicValueTotal) = _proofOutputs.get(0).extractProofOutput();
+
+        require(publicValueTotal == 0, "burn transactions cannot have a public value");
```

Comment:
Same comment as in `mint` - we should move these checks into the validator smart contract

Reply:
Good point, done